# Technical Challenge ! Custiomers and Orders - The lifeblood of any business!

This notebook depends on two data files: customers.scv and orders.csv. These files are hosted in the cloud (or Github), so for simplicity, we are providing you with the code to download and save the two files in a folder called 'sample_data' which lives within this notebook session (not on your local computer nor Google Drive).

![sample_data folder in notebook session](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/sample_data_folder.png)

Please execute the first code snippet before moving to the rest of the exercise as this code will import the data from the files for you to be able to execute the rest of the exercise.

Remember as we said in the previous lecture, you have to save this notebook into your Google Drive by going to menu File/Save a Copy In Drive, or clicking in the button "Copy to Drive". This way, you will not lose any work you've done and the file will persist with your latest changes in your Google Drive. Also we recommend you to rename your .ipynb in your Google Drive, so you can easily find it in the future.

In [1]:
import requests

def import_data_files():
  r = requests.get('https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/customers.csv')
  with open('./sample_data/customers.csv', 'wb') as f:
    f.write(r.content)

  r = requests.get('https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/orders.csv')
  with open('./sample_data/orders.csv', 'wb') as f:
    f.write(r.content)

import_data_files()
print("Customers and orders CSV files have been added './sample_data'")

Customers and orders CSV files have been added './sample_data'


# Exercise 1: Processing Customers data (difficulty medium)

The sample customer data in 'customers.csv' file has just 5 columns: CustomerId, First Name, Last Name, City and State

![Data sample](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/customers.png)

We strongly recommend that you complete from the [Prep Course: Intro to Python](https://colab.research.google.com/github/anyoneai/notebooks/blob/main/python3_crash_course.ipynb) the following section:
- Section 7: File I/O (to understand how to read CSV file)
- Section 6: For Loop (to navigate the contents of the CSV file)
- Section 5: Tuples, Lists, and Dictionaries (to manipulate the data of the CSV file)

With this, we hope you can complete this exercise successfully. Although if you want to solve this with libraries or any other way, you are welcome to do it your way.

*Hint:* We advise you to take a look at the data before you start.
**if you want to manually take a look at the data before you start, please see the content of the data [here](https://github.com/anyoneai/notebooks/blob/main/customers_and_orders/data/customers.csv).

*Hint*: There are many ways to do this exercise you can do your own, although here's some help. You can solve this exercise reading and parsing CSV files, structuring data into dictionaries, and using for loops to navigate the contents

*Hint*: Also, keep in mind that data might not be clean and you might have to figure out how to deal with that data from the code, without having to modify the data source.

Below are the 5 questions you'll have to answer to pass the evaluation:

**Question 1:** How many customers are in the file?
(as help, we have added some comments and starter code to help you structure the solution)

In [3]:
import numpy as np
import pandas as pd

ruta = "/content/sample_data/customers.csv"

customer_data = pd.read_csv(ruta)

customer_data

CustomerID    FirstName      LastName         City State
0          6192        Randi    Piedrahita  Canoga Park    CA
1       5100595  Christopher       Abraham   Loma Linda    CA
2       1902451        Elvis      Jauregui  Los Angeles    CA
3       1906377         Jose  Mota-Sanchez       Downey    CA
4       1902406      Scott        McCrorie   Long Beach    CA
..          ...          ...           ...          ...   ...
597     5415365      Trinity     Hall Reif        Poway    CA
598     5415465       Kisona   Mialoto Jr.      Lynwood    CA
599     5415345     Lawrence         Zhang  Temple City    CA
600     5415481         John        Hughes    Round Top    TX
601     5415482      Phillip        Okonma  Los Angeles    CA

[602 rows x 5 columns]

In [7]:
customer_data.shape

(602, 5)

In [6]:
customer_data["CustomerID"].nunique()

471

In [17]:
customer_data.sort_values(by="CustomerID", ascending=True)

CustomerID FirstName     LastName                       City  \
141        5014    Teresa     Ascolese              Santa Barbara   
44         5053       Doe       Harris                Los Angeles   
95         5053       Doe       Harris  Canoga Park                 
94         5053       Doe       Harris            North Hollywood   
443        5068   Maureen  Amato Mayes              Santa Barbara   
..          ...       ...          ...                        ...   
567     5415498     Ayman      Youssef                   Monrovia   
568     5415499  Mariella         Rudi                Los Angeles   
572     5415500     Volha    Kirilenko                  San Diego   
133     5415502  Geovanny     Espinoza                  Claremont   
534     5415503    Alison       Rivera                 South Gate   

             State  
141             CA  
44              CA  
95   CA             
94              CA  
443             CA  
..             ...  
567             CA  
568             CA  
572             CA  
133             CA  
534             CA  

[602 rows x 5 columns]

In [18]:
customer_data["CustomerID"].value_counts()

CustomerID
5463       5
10986      5
1909384    4
4301117    4
700771     4
          ..
5370647    1
5370268    1
5370211    1
5371053    1
5173569    1
Name: count, Length: 471, dtype: int64

### **ANSWER**
As we can see from the analysis made, there are 602 rows of data from customers but `there are 471 unique customers`. There are customers that have more than 1 appearance which can be misleading.

### **RESPUESTA**
Como podemos ver en el análisis realizado, hay 602 filas de datos de clientes, pero `la cantidad de clientes únicos son 471`, ya que hay clientes que tienen más de una instancia en el dataset, lo que puede llevar a pensar que son 602.

**Question 2:** In how many different states do the customers live in?

In [20]:
display(customer_data["State"].nunique())

display(customer_data["State"].unique())

16

array(['CA', 'AZ', 'NV', 'FL', 'CA           ', 'WA', 'NH', 'ID', 'CO',
       'Ca', 'TX', 'NM', 'OR', 'UT', 'MA', 'IN'], dtype=object)

In [25]:
states = customer_data["State"].unique()

states_clean = []

for state in states:
    clean_state = state.lower().strip()
    if clean_state not in states_clean:
        states_clean.append(clean_state)

display(states_clean)
display(len(states_clean))

['ca',
 'az',
 'nv',
 'fl',
 'wa',
 'nh',
 'id',
 'co',
 'tx',
 'nm',
 'or',
 'ut',
 'ma',
 'in']

14

### **ANSWER**
At first glance with the .nunique function we can observe 16 unique/different states but when we observe this values the are some inconsistencies:
- CA is different from Ca and (CA&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;) but they all are referring to California.
- After cleaning this mess we can see `there are 14 unique/different states where customers live in`.

### **RESUMEN**
A primera vista, cuando utilizamos la función .nunique podemos ver que hay 16 diferentes estados donde los clientes residen, pero cuando observamos los valores hay algunas inconsistencias:
- CA es diferente a Ca y (CA&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;) pero todos estos valores hacen referencia a California.
- Despúes de limpiar estos valores obtenemos que `hay 14 estados diferentes donde residen los clientes`.

**Question 3** What is the state with most customers?

In [ ]:
# open file and init vars
# evaluate the contents of the file

**Question 4** What is the state with the least customers?

In [ ]:
# open file and init vars
# evaluate the contents of the file

**Question 5:** What is the most common last name?

In [ ]:
# open file and init vars
# evaluate the contents of the file

# Exercise 2: Processing Orders data (difficulty high)

The second sample files contains orders placed by customers from the first file. Be careful, this file has many rows and you most likely should not print the contents of the whole file.

The file contains the following columns: CustomerID, OrderID, Date, OrderTotal, ProductName, Price

![Data sample](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/images/orders.png)

*Hint:* We advise you to take a look at the data before you start.
**if you want to manually take a look at the data before you start, please see the content of the data [here](https://raw.githubusercontent.com/anyoneai/notebooks/main/customers_and_orders/data/orders.csv).

*Hint*: There are many ways to do this exercise you can do your own, although here's some help. You can solve this exercise reading and parsing CSV files, structuring data into dictionaries, and using for loops to navigate the contents

*Hint*: Also, the data is not clean and you will have to figure out how to deal with that data from the code, without having to modify the data source.



**Question #1:** How many unique orders are in the orders.csv file?

**Question #2:** What is the average number of items per order (rounded to two decimal places)?

**Question #3:** What is the highest number of items per order?

**Question #4:** What is the number of orders placed in October 2021?

**Question #5:** Which customer spent the most amount of money in 2021?

**Question #6:** Historically, what is the best month for sales?

Once you get your answers, remember to go back to the course and introduce them in the multiple choice quiz

In [ ]:
# open and evaluate orders.csv

# Finished!

Hope this was not too difficult and slicing and dicing the datasets was some fun. Now head on back to the course and provide the answers to the questions from this exercise.